### DISCUSSION TITLE
Best Practices for Setting Up DevContainers for Both Windows and Linux

### DISCUSSION THREAD

**Discussion by MightyMirko** [devcontainers/community#149] [Q&A]
Hello everyone,

I'm looking for advice on how to configure development containers (DevContainers) within a single repository so that they are ready to run seamlessly on both Windows and Linux environments.

**Context:**

I have a repository where I want to use development containers to standardize the development environment for different projects. My goal is to ensure that these containers work efficiently on both Windows and Linux machines.

**Questions:**

1. **Multi-Platform Support:** What are the best practices for configuring DevContainers to ensure compatibility with both Windows and Linux? Are there specific configurations or tools that can help streamline this process?

2. **Conditional Arguments:** Is there a recommended approach to conditionally apply certain Docker or Podman arguments based on the host operating system? For example, Podman requires specific flags that Docker does not support.

3. **Script Execution:** How can I effectively manage the execution of setup scripts or commands that might differ between Windows and Linux? Is it advisable to have separate scripts for each OS, or can a single script be made to handle both?

4. **Environment Variables:** What strategies can be used to manage environment variables that might need to be set differently based on the host OS? 

5. **Testing and Validation:** How should I test and validate that my DevContainer configuration works correctly across different platforms?

**Current Setup:**

Here's a brief overview of my current setup:
- A `devcontainer.json` file to define the DevContainer configuration.
- A Dockerfile for building the container image.
- Potential use of Docker Compose for additional container management.

**Example Configuration:**

**devcontainer.json:**
```json
{
    "name": "Custom Development Container",
    "build": {
        "dockerfile": "Dockerfile"
    },
    "updateRemoteUserUID": false,
    "remoteUser": "vscode",
    "customizations": {
        "vscode": {
            "extensions": [
                "ms-vscode.cpptools",
                "ms-vscode.cmake-tools"
            ]
        }
    },
    "onCreateCommand": "bash .devcontainer/setup.sh"
}
```

**Dockerfile:**
```Dockerfile
FROM ubuntu:22.04
# Install necessary packages
RUN apt-get update && apt-get install -y \
    build-essential \
    cmake \
    git
```

**Questions:**
1. How can I configure `setup.sh` to handle OS-specific logic? How can it run on Windows?
2. How can I dynamically determine if Docker or Podman is used and adjust settings accordingly?
3. How can I dynamically determine if Windows or Linux is used and adjust settings accordingly?
4. Is there a way to template the devcontainer.json based on environment var such as chezmoi dotfiles?

Any advice, examples, or resources you could share would be greatly appreciated!

Thank you!

**Comment by MightyMirko** [comment]
Initially, I encountered an issue with using the --userns=keep-id argument, which is specific to Podman and not supported by Docker. When configuring the DevContainer, I needed a way to apply this argument only when Podman is used, while ensuring that Docker configurations remain unaffected. This discrepancy led to difficulties in achieving a consistent development environment across different platforms. My challenge was to find an effective method to handle this argument conditionally within a single configuration setup.

Another challenge I faced was managing file permissions within the container, especially in a way that is compatible with both Windows and Linux environments. File permissions and user management can vary significantly between these operating systems, and ensuring that the correct permissions are set automatically is crucial for a smooth development experience. Ideally, I would like to automate the configuration of these permissions based on the host operating system. This includes setting appropriate user IDs, group IDs, and handling any necessary file ownership adjustments. How can I best automate this process in the DevContainer setup to ensure consistent permissions and avoid issues related to access rights across different platforms?

**Comment by samruddhikhandale** [comment]
Hi 👋 

To configure development containers (Dev Containers) that work seamlessly on both Windows and Linux environments, you can follow these best practices and strategies:

1. Multi-Platform Support

- Use Cross-Platform Base Images: Ensure your Dockerfile uses a base image that works on both Windows and Linux.
- Volume Mounting: Be cautious with volume mounting paths as they differ between Windows and Linux. Use relative paths where possible.

2. Conditional Arguments

- Docker vs. Podman: You can use environment variables or scripts to detect the container runtime and apply specific arguments. For example, you can check for the presence of podman or docker in the system PATH.

3. Script Execution

- Cross-Platform Scripts: Use scripts that can handle both Windows and Linux. You can write scripts in a cross-platform language like Python or use shell scripts with conditional logic.
- Separate Scripts: Alternatively, maintain separate scripts for Windows and Linux and invoke the appropriate one based on the detected OS.

4. Environment Variables

- Conditional Environment Variables: Use environment variables to set different values based on the host OS. You can set these variables in your devcontainer.json or Dockerfile.

5. Testing and Validation

- Automated Testing: Use CI/CD pipelines to test your DevContainer configuration on both Windows and Linux environments.
- Manual Testing: Manually test the setup on both platforms to ensure compatibility.

### Example Configuration

#### devcontainer.json:

```jsonc
{
    "name": "Custom Development Container",
    "build": {
        "dockerfile": "Dockerfile"
    },
    "updateRemoteUserUID": false,
    "remoteUser": "vscode",
    "customizations": {
        "vscode": {
            "extensions": [
                "ms-vscode.cpptools",
                "ms-vscode.cmake-tools"
            ]
        }
    },
    "onCreateCommand": "bash .devcontainer/setup.sh"
}
```

#### Dockerfile

```
FROM mcr.microsoft.com/devcontainers/base:1-ubuntu-22.04

# Install necessary packages
RUN apt-get update && apt-get install -y \
    build-essential \
    cmake \
    git

# Copy setup script
COPY .devcontainer/setup.sh /setup.sh
RUN chmod +x /setup.sh
```

#### setup.sh

```
#!/bin/bash

# Detect OS
OS="$(uname -s)"
case "${OS}" in
    Linux*)     machine=Linux;;
    Darwin*)    machine=Mac;;
    CYGWIN*)    machine=Cygwin;;
    MINGW*)     machine=MinGw;;
    *)          machine="UNKNOWN:${OS}"
esac

echo "Detected OS: ${machine}"

# OS-specific logic
if [[ "$machine" == "Linux" ]]; then
    echo "Running Linux-specific setup"
    # Linux-specific commands
elif [[ "$machine" == "MinGw" ]]; then
    echo "Running Windows-specific setup"
    # Windows-specific commands
else
    echo "Unsupported OS"
    exit 1
fi
```

#### Detecting Docker or Podman:

```
if command -v podman &> /dev/null
then
    echo "Podman detected"
    # Podman-specific commands
elif command -v docker &> /dev/null
then
    echo "Docker detected"
    # Docker-specific commands
else
    echo "No container runtime detected"
    exit 1
fi
```

By following these strategies, you can create a dev container setup that works seamlessly across both Windows and Linux environments. This approach ensures that your development environment is consistent and reliable, regardless of the underlying OS.

**Reply by Adjective-Object** [reply]
I'm also running into this issue -- specifically, there doesn't seem to be a way to make `runArgs` conditional on the configured `dev.containers.dockerPath` setting. I have a few questions about your proposed solutions:

> Conditional Arguments
> Docker vs. Podman: You can use environment variables or scripts to detect the container runtime and apply specific arguments. For example, you can check for the presence of podman or docker in the system PATH.

It is possible to detect if a program is in a PATH directory purely from the env? I believe you would need to shell out to `which podman`, which AFAIK isn't permitted in the devcontainer config.

Your proposed solution also seems like it's built around running commands during the docker build process, which isn't the root of the problem. We need to change the arguments that `podman` is launched with.

It seems like this simply isn't supported by vscode, which is unfortunate.

**Comment by westurner** [comment]
For podman rootless containers, I need this configuration:
```json
	"runArgs": [
		"--gpus", "all",
		"--userns=keep-id",
		"--security-opt=label=disable"
	],
```

But for Docker (and GitHub Codespaces), `--userns=keep-id` doesn't work, so it needs to be this:

```json
	"runArgs": [
		"--gpus","all",
	],
```

Workarounds:

- Shim shell script:
  - Create a shell script shim that adds `--userns=keep-id` when the command is `run`
  - settings.json: `"dev.containers.dockerPath": "path/to/podman.sh"`

Solutions:

- Extend devcontainers spec
  - support runArgs_podman; such that runArgs + runArgs[dockerPath.basename]
  - IIRC there's an issue about this issue already

Usability would be better if it weren't necessary to create a `podman.sh` script so that the same devcontainer.json will work: (1) with podman rootless containers locally and; (2) with docker containers in github codespaces

**Reply by westurner** [reply]
This seems to work; such that I can omit the `runArgs` from `devcontainer.json`:

```sh
#!/bin/sh
# podman-rootless -- Call through to podman and add PODMAN_RUN_ARGS if cmd=run

# Usage:
# 1. Include podman-rootless on $PATH:
#    sudo cp ./podman-rootless /usr/local/bin/
#    ln -s ./podman-rootless ~/.local/bin/
#
# 2. In vscode settings.json:
#    "dev.containers.dockerPath": "/home/user/.local/bin/podman-rootless"

if [ -n "${DEBUG_PODMAN_ROOTLESS}" ]; then
    echo "${@}" >> ~/podman-rootless.log
fi

export PODMAN_RUN_ARGS=${PODMAN_RUN_ARGS:-"--userns=keep-id --security-opt=label=disable"}

cmd=$1
if [ "${cmd}" == "run" ] && [ ! -n "${SKIP_PODMAN_ROOTLESS}" ]; then
    shift
    podman run ${PODMAN_RUN_ARGS} "${@}"
else
    podman "${@}"
fi
```
